helix-view/src/gutter.rs

```diff
@@ -91,24 +126,41 @@ pub fn line_numbers<'doc>(
             } else {
                 line + 1
             };
+
+            let n_digits = count_digits(display_num);
+
             let style = if selected && is_focused {
                 linenr_select
             } else {
                 linenr
             };
-            write!(out, "{:>1$}", display_num, width).unwrap();
+
+            // if line number overflows maximum alotted width, truncate start
+            if n_digits > width {
+                let display_trailing = (display_num as u32) % 10_u32.pow((width - 1) as u32);
+                write!(out, "{}{:0>2$}", ELLIPSIS, display_trailing, width - 1).unwrap();
+            } else {
+                write!(out, "{:>1$}", display_num, width).unwrap();
+            }
+
             Some(style)
         }
     })
 }
 
+pub fn line_numbers_width(view: &View) -> usize {
+    // TODO: allow gutter widths to be dependent on Document. Currently the
+    // width is based on full View height, not visible line numbers.
+    let last_view_line = view.offset.row + view.area.bottom() as usize;
+    std::cmp::max(std::cmp::min(count_digits(last_view_line), 5), 3)
+}
```

Comment:
It would be great to have `Document` available here to base this line number width on `min(<last_document_line>, <last_visible_line>)`, but I quickly ended up with cyclic data flows. Total gutter width is needed for `view.inner_area` is needed for last line is needed for total gutter width.